File: lc-python/src/lc_0020_valid_parentheses.py
import unittest


def is_valid(s: str) -> bool:
    """
    Keep track of the opening braces in a list/stack. When a closing brace is
    found, check against the most recent opening brace.

    Time complexity: O(n)
    Space complexity: O(n)
    """
    if len(s) % 2 == 1:
        return False
    tokens = {"(": ")", "{": "}", "[": "]"}
    stack = []
    for ch in s:
        if ch in "({[":
            stack.append(ch)
        else:
            if len(stack) == 0 or tokens[stack.pop()] != ch:
                return False
    return len(stack) == 0


class TestValidParentheses(unittest.TestCase):

    def test_valid_parentheses(self):
        test_cases = [
            ["()", True],
            ["()[]{}", True],
            ["(]", False],
            ["[", False],
            ["]", False],
        ]
        for tc in test_cases:
            actual = is_valid(tc[0])
            expected = tc[1]
            self.assertEqual(actual, expected)


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/solutions/lc_0020_valid_parentheses.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";

function is_valid(s: string): boolean {
  if (s.length % 2 !== 0) {
    return false;
  }

  const open_brackets = "({[";
  const matching_brackets: { [key: string]: string } = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  const stack: string[] = [];
  for (let i = 0; i < s.length; i += 1) {
    if (open_brackets.includes(s[i])) {
      stack.push(s[i]);
      continue;
    }

    if (stack.length === 0) {
      return false;
    }

    const ch = stack.pop()!;
    if (s[i] !== matching_brackets[ch]) {
      return false;
    }
  }

  return stack.length === 0;
}

Deno.test("0020 Valid Parenthesis", () => {
  const test_cases: [string, boolean][] = [
    ["()", true],
    ["()[]{}", true],
    ["(]", false],
    ["[", false],
    ["]", false],
  ];

  for (let i = 0; i < test_cases.length; i += 1) {
    const test_case = test_cases[i];

    const str = test_case[0];
    const expected = test_case[1];

    const actual = is_valid(str);

    assertEquals(actual, expected);
  }
});
